parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+                universe of possible candidates is very small.
+            * vocab: the set of all tokens in the vocabulary
+
+        TODO: Consider allowing examples in a batch to have different candidate
+        sets (requires changes to MemNN module). Currently, all examples in a
+        batch must have the same candidate set.
+        """
+        label_vecs = batch.label_vec # [bsz] list of lists of LongTensors
+        batchsize = batch.text_vec.shape[0]
+
+        if label_vecs is not None:
+            assert label_vecs.dim() == 2
+        label_inds = None
+
+        if source == 'batch':
+            self._warn_once(
+                f'{mode}_batch_candidates',
+                f'[ Executing {mode} mode with batch labels as set of '
+                    'candidates. ]')
+            if batchsize == 1:
+                self._warn_once(
+                    f'{mode}_batchsize_1',
+                    f"[ Warning: using candidate source 'batch' and observed a "
+                    "batch of size 1. This may be due to uneven batch sizes at "
+                    "the end of an epoch. ]")
+            if label_vecs is None:
+                raise ValueError("If using candidate source 'batch', then "
+                    "batch.label_vec can not be None.")
+
+            # Labels are single tokens, enforce uniqueness
+            if label_vecs.size(1) == 1:
+                cand_vecs, label_inds = label_vecs.unique(return_inverse=True)
+                cands = [batch.labels[ind] for ind in label_inds]
+            # To save computation, treat all label_vecs as unique
+            else:
+                cands = batch.labels
+                cand_vecs = label_vecs
+                label_inds = label_vecs.new(range(batchsize))
+
+        elif source == 'inline':
+            # WARNING: this is currently the slowest of the options
+            self._warn_once(
+                f'{mode}_batch_candidates',
+                f'[ Executing {mode} mode with provided inline set of candidates '
+                    '(assumed to be identical for all candidates in a batch). ]')
+            if batch.candidate_vecs is None:
+                raise ValueError("If using candidate source 'inline', then "
+                    "bath.candidate_vecs can not be None.")
+            if batchsize > 1 and (
+                torch.any(batch.candidate_vecs[0] != batch.candidate_vecs[1])):
+                raise ValueError("All examples in a batch must have the same "
+                    "candidate set.")
+
+            cands = batch.candidates[0]
+            cand_vecs, _ = padded_tensor(batch.candidate_vecs[0],
+                use_cuda=self.use_cuda)
+            if label_vecs is not None:
+                label_inds = label_vecs.new_empty(batchsize)
+                # Warning: This computation is O(batchsize x num_candidates x seqlen)
+                for i, label_vec in enumerate(label_vecs):
+                    # Option 1: matrix multiplies
```

Comment:
Comment seems weird? How is it matrix multiplies? Anyway, let's drop the option 1/option 2 comments and just leave the version you actually chose.